_posts/2020-06-30-sparklyr-1.3.0-released/sparklyr-1.3.0.Rmd

```diff
@@ -0,0 +1,195 @@
+```
+
+What would be even better than [sparklyr](https://sparklyr.ai) 1.2? Many brilliant minds from RStudio, Databricks, and the broader open-source community have contemplated this question for months. Now they have an answer -- sparklyr 1.3 is on CRAN today!
+
+This new version adds the following major features:
+
+ * [Higher-order Functions](#higher-order-functions) to easily manipulate arrays and structs.
+ * Support for Apache [Avro](#avro), a row-oriented data serialization framework.
+ * [Custom Serialization](#custom-serialization) using R functions to read and write any data format.
+
+To install `sparklyr` 1.3 from CRAN run,
+
+```r
+install.packages("sparklyr")
+```
+
+In this post, we shall highlight some major new features introduced in sparklyr 1.3 and showcase scenarios where such features come in handy. While a number of enhancements and bug fixes (especially those related to `spark_apply`, [Apache Arrow](https://arrow.apache.org/), and secondary Spark connections) were also an important part of this release, they will not be the topic of this post, and it will be an easy exercise for the reader to find out more about them from the sparklyr [NEWS](https://github.com/sparklyr/sparklyr/blob/master/NEWS.md) file.
+
+
+## Higher-order Functions
+
+[Higher order functions](https://issues.apache.org/jira/browse/SPARK-19480) are built-in Spark SQL constructs that allow user-defined lambda expressions to be applied efficiently to complex data types such as arrays and structs. As a quick demo to see why higher-order functions are useful, let's say one day Scrooge Mcduck dived into his huge vault of money and found large quantities of pennies, nickels, dimes, and quarters. Having an impeccable taste in data structure, he decided to store the quantities and face values of everything into 2 Spark SQL array columns:
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "2.4")
+coins_tbl <- copy_to(
+  sc,
+  tibble::tibble(
+    quantities = list(c(4000, 3000, 2000, 1000)),
+    values = list(c(1, 5, 10, 25))
+  )
+)
+```
+
+thus declaring his net worth of 4k pennies, 3k nickels, 2k dimes, and 1k quarters. To help Scrooge Mcduck calculate the total value of each type of coin in Spark 2.4 or higher, we can apply `hof_zip_with`, the sparklyr equivalent of [ZIP_WITH](https://spark.apache.org/docs/latest/api/sql/index.html#zip_with), to `quantities` column and `values` column, combining pairs of elements from arrays in both columns. As you might have guessed, we also need to specify how to combine those elements, and what better way to accomplish that than a concise one-sided formula `~ .x * .y` in R, which says we want (quantity * value) for each type of coin? So, we have the following:
+
+```{r echo=TRUE, eval=FALSE}
+result_tbl <- coins_tbl %>%
+  hof_zip_with(~ .x * .y, dest_col = total_values) %>%
+  dplyr::select(total_values)
+
+result <- collect(result_tbl)
+result$total_values
```

Comment:
I think this would also work, `pull()` is usually prefered when retrieving a single column:

```suggestion
coins_tbl %>%
  hof_zip_with(~ .x * .y, dest_col = total_values) %>%
  dplyr::select(total_values)
  dplyr::pull(total_values)
```